File: no-auth/src/main/java/com/kjj/noauth/client/UserClient.java
package com.kjj.noauth.client;

import com.kjj.noauth.dto.user.JoinDto;
import com.kjj.noauth.dto.user.UserDto;
import com.kjj.noauth.exception.CantFindByUsernameException;
import com.kjj.noauth.util.HttpTools;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Repository;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

@Repository
@RequiredArgsConstructor
public class UserClient {
    private final RestTemplate restTemplate;
    private final HttpTools httpTools;
    @Value("${my.client.user.host}") private String userServerHost;

    private String getRequestURI(String path) {
        return userServerHost + path;
    }

    private String getRequestURI(String path, Map<String, String> params) {
        StringBuilder stringBuilder = new StringBuilder(userServerHost + path + '?');
        for (Map.Entry<String, String> entry : params.entrySet()) {
            stringBuilder.append(String.join("=", entry.getKey(), entry.getValue()));
            stringBuilder.append("&");
        }
        return stringBuilder.toString();
    }


    public UserDto findByUsername(String username) {
        String uri = getRequestURI("/v1/user/info/all", Map.of("username", username));
        UserDto userDto = restTemplate.getForObject(uri, UserDto.class);
        if (userDto == null) throw new CantFindByUsernameException("""
                해당 username을 가진 유저를 찾을 수 없습니다.
                username = """ + username);
        return userDto;
    }

    public Boolean existsByUsername(String username) {
        String uri = getRequestURI("/v1/user/exist", Map.of("username", username));
        return restTemplate.getForObject(uri, Boolean.class);
    }

    public Boolean join(JoinDto dto) {
        String uri = getRequestURI("/v1/user/join");
        return restTemplate.postForObject(uri, httpTools.getRequestEntityTypeJson(dto), Boolean.class);
    }

    public UserDto joinKeycloak(UserDto dto) {
        String uri = getRequestURI("/v1/user/join/keycloak");
        return restTemplate.postForObject(uri, httpTools.getRequestEntityTypeJson(dto), UserDto.class);
    }

    public Boolean withdraw(String username) {
        String uri = getRequestURI("/v1/user/withdraw", Map.of("username", username));
        return restTemplate.getForObject(uri, Boolean.class);
    }

    public Boolean checkPassword(String username, String password) {
        String uri = getRequestURI("/v1/user/check/password", Map.of("username", username, "password", password));
        return restTemplate.getForObject(uri, Boolean.class);
    }
}

File: no-auth/src/main/kotlin/com/kjj/noauth/client/UserClient.kt
package com.kjj.noauth.client

import com.kjj.noauth.dto.user.JoinOrLoginDto
import com.kjj.noauth.dto.user.UserDto
import com.kjj.noauth.exception.CantFindByUsernameException
import com.kjj.noauth.tool.HttpTool
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.net.URI

@Service
class UserClient(
    private val restTemplate: RestTemplate,
    private val httpTool: HttpTool,
    @Value("\${my.client.user.host}") private val userServerHost: String
) {
    private fun getRequestURI(path: String): URI {
        return URI(userServerHost + path)
    }

    private fun getRequestURI(path: String, params: Map<String, String>): URI {
        val stringBuilder = StringBuilder("$userServerHost$path?")
        for ((key, value) in params) {
            stringBuilder.append(java.lang.String.join("=", key, value))
            stringBuilder.append("&")
        }
        return URI(stringBuilder.toString())
    }

    fun findByUsername(username: String): UserDto {
        val uri = getRequestURI(
            "/v1/client/user/info/all",
            mapOf(Pair("username", username))
        )
        return restTemplate.getForObject(uri, UserDto::class.java)
            ?: throw CantFindByUsernameException("""
                해당 username을 가진 유저를 찾을 수 없습니다.
                username = $username
                """
            )
    }

    fun existsByUsername(username: String): Boolean {
        val uri = getRequestURI("/v1/client/user/check", mapOf(Pair("username", username)))
        return restTemplate.getForObject(uri, Boolean::class.java)!!
    }

    fun join(dto: JoinOrLoginDto): Boolean {
        val uri = getRequestURI("/v1/client/user/join")
        return restTemplate.postForObject(uri, httpTool.getRequestEntityTypeJson(dto), Boolean::class.java)!!
    }

    fun joinKeycloak(dto: UserDto): UserDto {
        val uri = getRequestURI("/v1/client/user/join/keycloak")
        return restTemplate.postForObject(uri, httpTool.getRequestEntityTypeJson(dto), UserDto::class.java)!!
    }

    fun withdraw(username: String): Boolean {
        val uri = getRequestURI("/v1/client/user/withdraw", mapOf(Pair("username", username)))
        return restTemplate.getForObject(uri, Boolean::class.java)!!
    }

    fun checkPassword(username: String, password: String): Boolean {
        val uri = getRequestURI(
            "/v1/client/user/check/password",
            mapOf(
                Pair("username", username),
                Pair("password", password))
        )
        return restTemplate.getForObject(uri, Boolean::class.java)!!
    }
}